mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraACLMapper.java

```diff
@@ -61,55 +65,87 @@
     }
 
     private final CassandraAsyncExecutor executor;
-    private final Session session;
     private final int maxRetry;
     private final CodeInjector codeInjector;
+    private final PreparedStatement insertStatement;
+    private final PreparedStatement conditionalInsertStatement;
+    private final PreparedStatement conditionalUpdateStatement;
+    private final PreparedStatement readStatement;
 
     public CassandraACLMapper(Session session, CassandraConfiguration cassandraConfiguration) {
         this(session, cassandraConfiguration, () -> {});
     }
 
     public CassandraACLMapper(Session session, CassandraConfiguration cassandraConfiguration, CodeInjector codeInjector) {
-        this.session = session;
         this.executor = new CassandraAsyncExecutor(session);
         this.maxRetry = cassandraConfiguration.getAclMaxRetry();
         this.codeInjector = codeInjector;
+        this.insertStatement = session.prepare(insertCqlBase());
+        this.conditionalInsertStatement = session.prepare(insertCqlBase().ifNotExists());
+        this.conditionalUpdateStatement = prepareConditionalUpdate(session);
+        this.readStatement = prepareReadStatement(session);
+    }
+
+    private PreparedStatement prepareConditionalUpdate(Session session) {
+        return session.prepare(
+            update(CassandraACLTable.TABLE_NAME)
+                .where(eq(CassandraACLTable.ID, bindMarker(CassandraACLTable.ID)))
+                .with(set(CassandraACLTable.ACL, bindMarker(CassandraACLTable.ACL)))
+                .and(set(CassandraACLTable.VERSION, bindMarker(CassandraACLTable.VERSION)))
+                .onlyIf(eq(CassandraACLTable.VERSION, bindMarker(OLD_VERSION))));
+    }
+
+    private PreparedStatement prepareReadStatement(Session session) {
+        return session.prepare(
+            select(CassandraACLTable.ACL, CassandraACLTable.VERSION)
+                .from(CassandraACLTable.TABLE_NAME)
+                .where(eq(CassandraMailboxTable.ID, bindMarker(CassandraACLTable.ID))));
+    }
+
+    private Insert insertCqlBase() {
+        return insertInto(CassandraACLTable.TABLE_NAME)
+            .value(CassandraACLTable.ID, bindMarker(CassandraACLTable.ID))
+            .value(CassandraACLTable.ACL, bindMarker(CassandraACLTable.ACL))
+            .value(CassandraACLTable.VERSION, INITIAL_VALUE);
```

Comment:
what happens when there is to successive insert? the second one is considered as replacing the first one?

Reply:
That is the expected semantic, yes. Concurrent updates with the semantic defined in jmap can not be well handled as such.